File: pushed_android/app/src/main/java/com/pushed/android/auth/AuthManager.kt
package com.pushed.android.auth

import android.content.Context
import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GoogleAuthProvider
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manager for Firebase Authentication.
 * 
 * Provides unified authentication across the Pushed system, ensuring
 * the same UID is used on both Android sender devices and watchOS receivers.
 * 
 * Supports:
 * - Google Sign-In
 * - Email/Password authentication
 */
@Singleton
class AuthManager @Inject constructor(
    @ApplicationContext private val context: Context,
    private val firebaseAuth: FirebaseAuth
) {

    /**
     * Observable flow of the current authentication state.
     */
    val authState: Flow<AuthState> = callbackFlow {
        val listener = FirebaseAuth.AuthStateListener { auth ->
            val state = auth.currentUser?.let { user ->
                AuthState.Authenticated(
                    userId = user.uid,
                    email = user.email,
                    displayName = user.displayName,
                    photoUrl = user.photoUrl?.toString()
                )
            } ?: AuthState.Unauthenticated

            trySend(state)
        }

        firebaseAuth.addAuthStateListener(listener)

        awaitClose {
            firebaseAuth.removeAuthStateListener(listener)
        }
    }

    /**
     * Get the current user if authenticated.
     */
    val currentUser: FirebaseUser?
        get() = firebaseAuth.currentUser

    /**
     * Get the current user ID if authenticated.
     */
    val currentUserId: String?
        get() = firebaseAuth.currentUser?.uid

    /**
     * Check if user is currently authenticated.
     */
    val isAuthenticated: Boolean
        get() = firebaseAuth.currentUser != null

    /**
     * Sign in with Google credential.
     * 
     * @param idToken The Google ID token from Google Sign-In
     * @return Result containing the authenticated user or error
     */
    suspend fun signInWithGoogle(idToken: String): Result<FirebaseUser> {
        return try {
            val credential = GoogleAuthProvider.getCredential(idToken, null)
            val authResult = firebaseAuth.signInWithCredential(credential).await()
            
            authResult.user?.let { user ->
                Log.i(TAG, "Google sign-in successful: ${user.uid}")
                Result.success(user)
            } ?: Result.failure(AuthException("Sign-in succeeded but user is null"))
        } catch (e: Exception) {
            Log.e(TAG, "Google sign-in failed", e)
            Result.failure(AuthException("Google sign-in failed: ${e.message}", e))
        }
    }

    /**
     * Sign in with email and password.
     * 
     * @param email User's email address
     * @param password User's password
     * @return Result containing the authenticated user or error
     */
    suspend fun signInWithEmail(email: String, password: String): Result<FirebaseUser> {
        return try {
            val authResult = firebaseAuth.signInWithEmailAndPassword(email, password).await()
            
            authResult.user?.let { user ->
                Log.i(TAG, "Email sign-in successful: ${user.uid}")
                Result.success(user)
            } ?: Result.failure(AuthException("Sign-in succeeded but user is null"))
        } catch (e: Exception) {
            Log.e(TAG, "Email sign-in failed", e)
            Result.failure(AuthException("Email sign-in failed: ${e.message}", e))
        }
    }

    /**
     * Create a new account with email and password.
     * 
     * @param email User's email address
     * @param password User's password
     * @return Result containing the created user or error
     */
    suspend fun createAccountWithEmail(email: String, password: String): Result<FirebaseUser> {
        return try {
            val authResult = firebaseAuth.createUserWithEmailAndPassword(email, password).await()
            
            authResult.user?.let { user ->
                Log.i(TAG, "Account created successfully: ${user.uid}")
                Result.success(user)
            } ?: Result.failure(AuthException("Account creation succeeded but user is null"))
        } catch (e: Exception) {
            Log.e(TAG, "Account creation failed", e)
            Result.failure(AuthException("Account creation failed: ${e.message}", e))
        }
    }

    /**
     * Send password reset email.
     * 
     * @param email User's email address
     * @return Result indicating success or error
     */
    suspend fun sendPasswordResetEmail(email: String): Result<Unit> {
        return try {
            firebaseAuth.sendPasswordResetEmail(email).await()
            Log.i(TAG, "Password reset email sent to $email")
            Result.success(Unit)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send password reset email", e)
            Result.failure(AuthException("Failed to send password reset: ${e.message}", e))
        }
    }

    /**
     * Sign out the current user.
     */
    fun signOut() {
        Log.i(TAG, "Signing out user: ${currentUserId}")
        firebaseAuth.signOut()
    }

    /**
     * Reload the current user's profile.
     * Useful after email verification or profile updates.
     */
    suspend fun reloadUser(): Result<FirebaseUser> {
        return try {
            currentUser?.reload()?.await()
            currentUser?.let { user ->
                Result.success(user)
            } ?: Result.failure(AuthException("No user to reload"))
        } catch (e: Exception) {
            Log.e(TAG, "Failed to reload user", e)
            Result.failure(AuthException("Failed to reload user: ${e.message}", e))
        }
    }

    companion object {
        private const val TAG = "AuthManager"
    }
}

/**
 * Represents the current authentication state.
 */
sealed class AuthState {
    data object Unauthenticated : AuthState()
    
    data class Authenticated(
        val userId: String,
        val email: String?,
        val displayName: String?,
        val photoUrl: String?
    ) : AuthState()
}

/**
 * Exception for authentication errors.
 */
class AuthException(
    message: String,
    cause: Throwable? = null
) : Exception(message, cause)

File: pushed_watch/pushed_watch/Auth/AuthManager.swift
import FirebaseAuth
import Foundation
import os.log

/// Logger for authentication
private let logger = Logger(subsystem: "com.pushed.watch", category: "Auth")

/// Manager for Firebase Authentication on watchOS.
///
/// Provides unified authentication across the Pushed system, ensuring
/// the same UID is used on both Android sender devices and watchOS receivers.
///
/// **Important**: Signing identity must be `adityagurjar.it18@gmail.com`.
/// Do NOT use `a.gurjar@scbank.com.eg`.
@MainActor
@Observable
final class AuthManager {

  // MARK: - State

  /// Current authentication state
  var authState: AuthState = .loading {
    didSet {
      logger.info(
        "🔐 authState changed: \(String(describing: oldValue)) -> \(String(describing: self.authState))"
      )
    }
  }

  /// Current user ID if authenticated
  var currentUserId: String? {
    let uid = Auth.auth().currentUser?.uid
    logger.debug("🔐 currentUserId accessed: \(uid ?? "nil")")
    return uid
  }

  /// Whether user is currently authenticated
  var isAuthenticated: Bool {
    let authenticated = Auth.auth().currentUser != nil
    logger.debug("🔐 isAuthenticated accessed: \(authenticated)")
    return authenticated
  }

  /// Current user's email
  var userEmail: String? {
    Auth.auth().currentUser?.email
  }

  /// Current user's display name
  var displayName: String? {
    Auth.auth().currentUser?.displayName
  }

  // MARK: - Private

  private nonisolated(unsafe) var authStateListener: AuthStateDidChangeListenerHandle?

  // MARK: - Initialization

  init() {
    logger.info("🔧 AuthManager initializing...")
    logger.info(
      "🔐 Firebase Auth.auth().currentUser on init: \(Auth.auth().currentUser?.uid ?? "nil")")
    setupAuthStateListener()
    logger.info("🔧 AuthManager initialized")
  }

  deinit {
    if let listener = authStateListener {
      Auth.auth().removeStateDidChangeListener(listener)
    }
  }

  // MARK: - Public Methods

  /// Sign in with email and password.
  ///
  /// - Parameters:
  ///   - email: User's email address
  ///   - password: User's password
  /// - Returns: The authenticated user's UID
  func signIn(email: String, password: String) async throws -> String {
    logger.info("🔐 signIn() called with email: \(email)")
    do {
      let result = try await Auth.auth().signIn(withEmail: email, password: password)
      logger.info("✅ signIn() successful - uid: \(result.user.uid)")
      logger.info("   email: \(result.user.email ?? "nil")")
      logger.info("   displayName: \(result.user.displayName ?? "nil")")
      return result.user.uid
    } catch {
      logger.error("❌ signIn() failed: \(error.localizedDescription)")
      throw AuthError.signInFailed(error.localizedDescription)
    }
  }

  /// Create a new account with email and password.
  ///
  /// - Parameters:
  ///   - email: User's email address
  ///   - password: User's password
  /// - Returns: The created user's UID
  func createAccount(email: String, password: String) async throws -> String {
    logger.info("🔐 createAccount() called with email: \(email)")
    do {
      let result = try await Auth.auth().createUser(withEmail: email, password: password)
      logger.info("✅ createAccount() successful - uid: \(result.user.uid)")
      return result.user.uid
    } catch {
      logger.error("❌ createAccount() failed: \(error.localizedDescription)")
      throw AuthError.accountCreationFailed(error.localizedDescription)
    }
  }

  /// Sign out the current user.
  func signOut() throws {
    logger.info("🔐 signOut() called")
    do {
      try Auth.auth().signOut()
      logger.info("✅ signOut() successful")
    } catch {
      logger.error("❌ signOut() failed: \(error.localizedDescription)")
      throw AuthError.signOutFailed(error.localizedDescription)
    }
  }

  /// Send password reset email.
  ///
  /// - Parameter email: User's email address
  func sendPasswordReset(to email: String) async throws {
    logger.info("🔐 sendPasswordReset() called for email: \(email)")
    do {
      try await Auth.auth().sendPasswordReset(withEmail: email)
      logger.info("✅ Password reset email sent")
    } catch {
      logger.error("❌ sendPasswordReset() failed: \(error.localizedDescription)")
      throw AuthError.passwordResetFailed(error.localizedDescription)
    }
  }

  /// Reload current user's profile.
  func reloadUser() async throws {
    logger.info("🔐 reloadUser() called")
    guard let user = Auth.auth().currentUser else {
      logger.error("❌ reloadUser() failed: No current user")
      throw AuthError.notAuthenticated
    }

    try await user.reload()
    updateAuthState()
    logger.info("✅ User reloaded successfully")
  }

  // MARK: - Private Methods

  private func setupAuthStateListener() {
    logger.info("🔐 Setting up auth state listener...")
    authStateListener = Auth.auth().addStateDidChangeListener { [weak self] _, user in
      logger.info("🔔 Auth state listener fired!")
      logger.info("   user: \(user?.uid ?? "nil")")
      logger.info("   email: \(user?.email ?? "nil")")
      self?.handleAuthStateChange(user: user)
    }
    logger.info("🔐 Auth state listener set up")
  }

  private func handleAuthStateChange(user: User?) {
    logger.info("🔐 handleAuthStateChange() called")
    logger.info("   user: \(user?.uid ?? "nil")")

    if let user {
      logger.info("✅ User is authenticated")
      logger.info("   uid: \(user.uid)")
      logger.info("   email: \(user.email ?? "nil")")
      logger.info("   displayName: \(user.displayName ?? "nil")")
      authState = .authenticated(
        userId: user.uid,
        email: user.email,
        displayName: user.displayName
      )
    } else {
      logger.info("⚠️ User is NOT authenticated")
      authState = .unauthenticated
    }
  }

  private func updateAuthState() {
    logger.info("🔐 updateAuthState() called")
    handleAuthStateChange(user: Auth.auth().currentUser)
  }
}

// MARK: - Auth State

enum AuthState: Equatable {
  case loading
  case unauthenticated
  case authenticated(userId: String, email: String?, displayName: String?)

  var isAuthenticated: Bool {
    if case .authenticated = self {
      return true
    }
    return false
  }
}

// MARK: - Errors

enum AuthError: LocalizedError {
  case notAuthenticated
  case signInFailed(String)
  case accountCreationFailed(String)
  case signOutFailed(String)
  case passwordResetFailed(String)

  var errorDescription: String? {
    switch self {
    case .notAuthenticated:
      "User is not authenticated"
    case .signInFailed(let reason):
      "Sign in failed: \(reason)"
    case .accountCreationFailed(let reason):
      "Account creation failed: \(reason)"
    case .signOutFailed(let reason):
      "Sign out failed: \(reason)"
    case .passwordResetFailed(let reason):
      "Password reset failed: \(reason)"
    }
  }
}
